cohttp-curl-lwt/src/cohttp_curl_lwt.ml

```diff
@@ -0,0 +1,247 @@
+module Sink = struct
+  type _ t = String : string t | Discard : unit t
+
+  let string = String
+  let discard = Discard
+end
+
+module Source = struct
+  type t = Empty | String of string
+
+  let empty = Empty
+  let string s = String s
+
+  let to_curl_callback t =
+    match t with
+    | Empty -> fun _ -> ""
+    | String s ->
+        let len = String.length s in
+        let pos = ref 0 in
+        fun max_asked ->
+          if !pos >= len then ""
+          else
+            let chunk_len = min (len - !pos) max_asked in
+            let res = String.sub s !pos chunk_len in
+            pos := !pos + chunk_len;
+            res
+end
+
+module Context = struct
+  type t = {
+    mt : Curl.Multi.mt;
+    wakeners : (Curl.t, Curl.curlCode Lwt.u) Hashtbl.t;
+    all_events : (Unix.file_descr, Lwt_engine.event list) Hashtbl.t;
+    mutable timer_event : Lwt_engine.event;
+  }
+
+  let create () =
+    (* Most of this is taken from https://github.com/ygrek/ocurl/blob/master/curl_lwt.ml *)
+    let t =
+      {
+        mt = Curl.Multi.create ();
+        wakeners = Hashtbl.create 32;
+        all_events = Hashtbl.create 32;
+        timer_event = Lwt_engine.fake_event;
+      }
+    in
+    let rec finished s =
+      match Curl.Multi.remove_finished t.mt with
+      | None -> ()
+      | Some (h, code) ->
+          (match Hashtbl.find_opt t.wakeners h with
+          | None -> Printf.eprintf "curl_lwt(%s): orphan handle" s
```

Comment:
is this a debug thing, or should it raise an assertion failure?

Reply:
I'm trying to reproduce when this happens. If I don't manage to, I will turn this into an assertion.

Reply:
It's gone.